coursera/api.py

```diff
@@ -455,6 +464,116 @@ def extract_links_from_exam(self, exam_id):
+                if os.path.isdir(self._course_name + "/notebook/" + head + "/") == False:
+                    logging.info('Creating [{}] directories...'.format(head))
+                    os.makedirs(self._course_name + "/notebook/" + head + "/")
+                
+                r = requests.get(tmpUrl.replace(" ", "%20"), cookies=self._session.cookies)
+                if os.path.exists(self._course_name + "/notebook/" + head + "/" + tail) == False:
+                    logging.info('Downloading {} into {}'.format(tail, head))
+                    with open(self._course_name + "/notebook/" + head + "/" + tail, 'wb+') as f:
+                        f.write(r.content)
+                else:
+                    logging.info('Skipping {}... (file exists)'.format(tail))
+
+
+                if not str(extension[1:]) in supplement_links:
+                    supplement_links[str(extension[1:])] = []
+                
+                supplement_links[str(extension[1:])].append((tmpUrl.replace(" ", "%20"), filename))
+
+
+            elif content['type'] == 'notebook':
+                tmpUrl = OPENCOURSE_NOTEBOOK_DOWNLOAD.format(path=content['path'], jupId=jupyterId, timestamp=int(time.time()))
+                filename, extension = os.path.splitext(clean_url(tmpUrl))
+                
+                head, tail = os.path.split(content['path'])
+                
+                if os.path.isdir(self._course_name + "/notebook/" + head + "/") == False:
+                    logging.info('Creating [{}] directories...'.format(head))
+                    os.makedirs(self._course_name + "/notebook/" + head + "/")
+                
+                r = requests.get(tmpUrl.replace(" ", "%20"), cookies=self._session.cookies)
+                if os.path.exists(self._course_name + "/notebook/" + head + "/" + tail) == False:
+                    logging.info('Downloading Jupyter {} into {}'.format(tail, head))
+                    with open(self._course_name + "/notebook/" + head + "/" + tail, 'wb+') as f:
+                        f.write(r.content)
+                else:
+                    logging.info('Skipping {}... (file exists)'.format(tail))
+
+                if not "ipynb" in supplement_links:
+                    supplement_links["ipynb"] = []
+                
+                supplement_links["ipynb"].append((tmpUrl.replace(" ", "%20"), filename))
+
+            else:
+                logging.info('Unsupported typename {} in notebook'.format(content['type']))
+                
+        return supplement_links
+
+
+    def _get_notebook_json(self, notebook_id, authorizationId):
+        
+        import re, time
+        headers = self._auth_headers_with_json()
+        reply = get_page(
+            self._session,
+            OPENCOURSE_NOTEBOOK_DESCRIPTIONS,
+            json=False,
+            authId=authorizationId,
+            headers=headers
+        )
+
+        jupyterId = re.findall(r"\"\/user\/(.*)\/tree\"", reply)
```

Comment:
I suggest putting this regular expression into the beginning of the module, using re.compile and giving it a proper var name.